src/main.c

```diff
@@ -73,6 +73,85 @@ static void help() {
             "Report bugs to: https://github.com/kspalaiologos/bzip3\n");
 }
 
+static void xwrite(const void * data, size_t size, size_t len, FILE * des) {
+    if (fwrite(data, size, len, des) != len) {
+        fprintf(stderr, "Write error: %s\n", strerror(errno));
+        exit(1);
+    }
+}
+
+static int read_errcheck(void * data, size_t size, size_t len, FILE * des, size_t * out) {
+    *out = fread(data, size, len, des);
+    if (ferror(des)) {
+        fprintf(stderr, "Read error: %s\n", strerror(errno));
+        return -1;
+    }
+    return 0;
+}
+
+/* 0: expected eof
+ * 1: read succeeded
+ * -1: unexpected eof or error
+ */
+static int read_eofcheck(void * data, size_t size, size_t len, FILE * des) {
+    size_t val = fread(data, size, len, des);
+    if (!val)
+        return 0;
+    if (val == len)
+        return 1;
+    if (feof (des)) {
+        fprintf(stderr, "Error: Corrupt file\n");
+        return -1;
+    }
+    fprintf(stderr, "Read error: %s\n", strerror(errno));
+    return -1;
+}
+
+static int read_fullcheck(void * data, size_t size, size_t len, FILE * des) {
+    switch (read_eofcheck (data, size, len, des)) {
+        case 0:
+            fprintf(stderr, "Error: Corrupt file\n");
+        default:
+            return -1;
+        case 1:
+            return 0;
+    }
+}
+
+static void close_out_file(FILE * des) {
+    if (des) {
+        int outfd = fileno(des);
+        int status;
+
+        if (fflush(des)) {
+            fprintf(stderr, "Error: Failed on fflush: %s\n", strerror(errno));
+            exit(1);
+        }
+
+        /* would have to use || outfd != -1 && !FlushFileBuffers(_get_osfhandle(outfd)) and then use GetLastError + FormatMessage(A?) */
+#ifndef __MSVCRT__
+        while (1) {
+            status = fsync(outfd);
+            if (status == -1) {
+                if (errno == EINVAL)
+                    break;
+                if (errno == EINTR)
+                    continue;
+                fprintf(stderr, "Error: Failed on fsync: %s\n", strerror(errno));
+                exit(1);
```

Comment:
If we exit with code 1 here, then why not exit with code 1 everywhere in the helper functions? This would greatly simplify the control flow.